MVP guidelines for Multiline box (#2599)
coveo/plasma: packages/website/docs/MultilineBox.md

> # Best Practices

Reviewer: Add a new line after each heading please

```suggestion
# Best Practices

```

> The Multiline box allows users to **add the same sequence of inputs multiple times**.  

Reviewer 2: @kpoulincoveo Before I review this, I would need to clarify the terminology we use. This is the component that allows users to enter multiple URLs when creating a web source, right?

```suggestion
The multiline box allows users to **provide several inputs of the same type**.  
```
or 
`The multiline box allows users to **provide a series of inputs of the same type**.`
Not sure what we mean with "sequence"... 🤔

Reviewer 3: I Like the second option better personally :eyes:

Author: @jlevesquecoveo The multiline also serves to build complex conditions (ex: Filter expression) 

By "sequence", I mean that each line can hold multiple inputs (ex: first name + last name + email). That bundle of different input is a sequence. And each of line can be repeated multiple times (ex: to provide a list of people)

In some cases, the sequence of inputs can vary from one line to another (although it's not recommended to do it but some case allows it such as condition. That means that the first line could have a sequence of (field+comparator+single select) but the second line could have only (field+comparator+numerical input)

I let you decide which version of what you propose works best.

Reviewer 2: OK, thanks for these details, @kpoulincoveo! Can the sequence contain only one input, though? Like this:
<img width="510" alt="image" src="https://user-images.githubusercontent.com/47796194/160683506-48711d4f-a2eb-42cf-a04f-4d58b4819e50.png">
Or is this a different component?

Author: Yes in that case it can (weither or not you use Multiline depend on the type of input, for instance, having a single select only on each line should be using a multi-select instead)

Each sequence appears on a different line, making it easier to consult.  

Do not use the multiline box to regroup inputs that have no connection to one another.   
Multiline works best when the sequence of inputs is identical on each line.  

Multiline box can combine multiple inputs together to form a more complex input.  
They can also combine text elements when an input or sequence requires more explanation.  

Only use the drag and drop function if the order of each sequence is important.   
When doing so, the list must always be presented to the user in the order they have set.

# Labeling
Keep titles short, preferably **under three words**.  

If the list of inputs is very simple, provide a descriptive title without action verbs.  
For example, write "Email list" rather than "Provide a list of emails”.

If the sequence of inputs is more complex, consider using a title with a trigger verb to clearly explain that the outcome will apply or consider the entire list.  
For example, write “Show content that matches" instead of "Filter rules”. 

> When using a dialog approach, always consider other inputs in the form so that it is cohesive with other inputs.

Reviewer 2: ```suggestion
You may choose to use a dialog approach with the labels. For example, you could have: "Among all items matching... [condition1] Show only the items with... [condition2]". In such case, consider the other inputs in the form so that your approach is coherent with all inputs.
```

# Help Text and Instructions
Avoid the need for help text or instruction for each individual input that appears in a multiline line. 

Instead, provide short instructions, above the lines, that will briefly explain the type of inputs to be provided in the input sequence.

# Feedback and Validation

> Make sure that a first sequence is **completed and valid** before allowing the addition of a new line. 

Reviewer 2: ```suggestion
Ensure that all inputs on the first line are **complete and valid** before allowing the addition of a new line. 
```

> If the multiline is optional, do not show an empty first line. Users should be required to add it manually if they need it. 

Reviewer: This is not a behaviour the component currently implements AFAIK

Reviewer 4: The component indeed renders a new "empty" box by default when it mounts, it is in fact problematic for my current work with the `queryExpressionBuilder` and I though about adding some `doNotAddAnEmptyBoxByDefaultBecauseIDontWantOne` prop down the road.

Reviewer 2: ```suggestion
If the multiline box is optional, don't show an empty first line. Users should add it manually if they need it. 
```

In the same fashion, if the multiline is mandatory, the first line should be visible and the user should not be allowed to delete it. 

Always let the users **add a new line manually** instead of automating it. This avoids empty lines when the form is submitted. 

If each line has a sequence of inputs of variable width, make sure lines width at least allows for delete buttons to appear vertically aligned.

# Related Components

If your use case doesn't match the guidelines above, consider using one of the following components instead:

> [Multiselect](#/form/multiselect) when the user only has to select options from a single list.

Reviewer: This link won't work. I think you need to use `https://plasma.coveo.com/form/MultiSelect`

> If your use case doesn't match the guidelines above, consider using one of the following components instead:
> [Multiselect](#/form/multiselect) when the user only has to select options from a single list.

Reviewer 2: ```suggestion
When the user must select one or more options from a single list, consider using a [multi select](https://plasma.coveo.com/form/MultiSelect) instead.
```

> If users don’t need to repeat inputs, use a custom form layout instead.

Reviewer 2: @kpoulincoveo I don't understand this, can you explain plz? 🙂

Reviewer 2: ```suggestion
Don't use a multiline box to group inputs that have no connection to one another. Use a custom form layout instead. For example, if you want a user to provide their name and email address, just provide two text inputs. However, if you want them to provide the name and email address of several people, use a multiline box and show the name and email inputs on one line for each person.
```

# Best Practices

A multiline box allows users to **add multiple inputs or multiple series of inputs, each on a different line**.

Inputs can be combined into a series to form a more complex input.  
Text elements can also be included when an input or series requires more explanation.

A multiline box works best when the sequence of inputs is identical on all lines. 
Don't use the multiline box to group inputs that have no connection to one another.    

Only offer the drag and drop feature when the order of the inputs or series is important.
In such case, the inputs/series must always be presented in the order set by the user.

# Labeling

Keep titles short, preferably **under three words**.  

If the list of inputs is very simple, provide a descriptive title without action verbs.  
For example, write "Email list" rather than "Provide a list of emails”.

> If the series of inputs are complex, consider using a title with an action verb to clarify the outcome of the choice to make.  

Reviewer 2: ```suggestion
If the series of inputs is complex, consider using a title with an action verb to clarify the outcome of the choice to make.  
```

> For example, write “Show content that matches" instead of "Filter rules”. 

Reviewer 2: ```suggestion
For example, write “Show content that matches..." instead of "Filter rules”. 
```
